### PR TITLE
checkpoint/restore: implement --manage-cgroups-mode ignore

### PR DESCRIPTION
**Pull request by kolyshkin**
This patchset fixes a few issues and adds support for `--manage-cgroups-mode ignore`. This option allows to restore a container into a different cgroup than the original one. A test case is added to check that it works as expected.

See individual commits for details.

Loosely based on https://github.com/opencontainers/runc/pull/3447 and my earlier work from 2021.

Closes: https://github.com/opencontainers/runc/pull/3447

Also, this uses the new mode in lazy checkpoint/restore test, hopefully fixing this test flakiness.

Fixes: https://github.com/opencontainers/runc/issues/2760
Fixes: https://github.com/opencontainers/runc/issues/2924
Fixes: https://github.com/opencontainers/runc/issues/2475

PS I am thinking whether this (`--manage-cgroups-mode ignore` should become a default in 1.2.0, because otherwise changing the `cgroupsPath` in container's `config.json` before restore doesn't have any effect)

### CODE REVIEW

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL (this, among the other things, fixes a few flaky tests we had for a few years)

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL

**Comment by kolyshkin** [comment]
@adrianreber PTAL

**Comment by 1235467Ad** [comment]
Keep me posted on the details.

**Comment by 1235467Ad** [comment]
Someone is pretending to be me

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL (this, among the other things, fixes a few flaky tests we had for a few years)

**Comment by kolyshkin** [comment]
@adrianreber PTAL

**Comment by adrianreber** [comment]
Looks good. Thanks!

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL 🙏🏻

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL 🙏🏻
This fixes a few real bugs on which I worked (on and off) for the last few years, and it's a pity to see this sitting 4 months without a review.

File: tests/integration/checkpoint.bats
Comment (AkihiroSuda) [inline tests/integration/checkpoint.bats:433]: I don't think this is robust, but I don't have an alternative idea, so LGTM.

Eventually we need to have a robust error reporting system.

**Review by AkihiroSuda — APPROVED** [review]

File: tests/integration/checkpoint.bats
Comment (kolyshkin) [reply tests/integration/checkpoint.bats:433]: This is just to print anything containing `Error` (with some context) in case the checkpointing failed. It's merely a way to debug criu failures, and does not affect the test itself (only its output in case of an error). The `|| true` part here is so that the test won't fail in case there's no error; IOW, we ignore grep exit code.

Alternatively, we could create an after-test artefact containing all the files, but this works OK so far.

Thinking about it, we might work on making runc do something like what grep does here in case of an error. Currently, this is just criu writing the log file.

**Comment by kolyshkin** [comment]
Rebased to resolve a conflict caused by #3655.

@AkihiroSuda please re-LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Review by mrunalp — APPROVED** [review]

File: tests/integration/checkpoint.bats
Comment (kolyshkin) [reply tests/integration/checkpoint.bats:433]: Filed https://github.com/opencontainers/runc/issues/3711 to not forget about it

**Pull request merged**
